README.md
teahouse
========

A small project to improve my Python skills


HISTORY.rst
.. :changelog:

History
-------

0.1.0 (2014-11)
---------------------

* First release on PyPI.


test_mymodule.py
import unittest
from mymodule import CoffeeMachine

teacup = CoffeeMachine.brew('tea')
coffee_cup = CoffeeMachine.brew('coffee')
bread_cup = CoffeeMachine.brew('bread')


def almost_equal(result_temp, expected_temp):    # tolerance for function slowness
    difference = expected_temp - result_temp
    return 0 <= difference <= 2


class TestMyModule(unittest.TestCase):

    def test_has_attribute(self):
        self.assertTrue(hasattr(teacup, 'full'))
        self.assertFalse(hasattr(teacup, 'fill'))

    def test_wrong_drink(self):
        self.assertIsInstance(bread_cup, str)
        self.assertFalse(hasattr(bread_cup, 'price'))

    def test_right_drink(self):
        self.assertIsInstance(teacup, object)
        self.assertTrue(hasattr(teacup, 'price'))
        self.assertEqual(teacup.price, 1.80)
        
    def test_drink(self):
        teacup.drink(0.2)
        self.assertRaises(Exception, teacup.drink, -2)
        self.assertEqual(0.8, teacup.full)

    def test_heat_up(self):
        self.assertRaises(Exception, teacup.heat_up, -20)
        self.assertTrue(teacup.temp < 80)
        temp = teacup.temp
        teacup.heat_up(10)
        self.assertTrue(almost_equal(teacup.temp, temp + 10))    

    def test_has_timestamp(self):
        self.assertTrue(hasattr(teacup, '_timestamp'))

    def test_cool_down(self):    
        self.assertEqual(coffee_cup._init_temp, 80)
        coffee_cup._timestamp -= 1200    # set back start time by 20 minutes
        start_temp = coffee_cup.temp
        coffee_cup.heat_up(20)
        self.assertTrue(almost_equal(coffee_cup.temp, start_temp + 20))


if __name__ == '__main__':
    unittest.main()

tests/test_teahouse.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_teahouse
----------------------------------

Tests for `teahouse` module.
"""

import unittest

from teahouse import CoffeeMachine


def almost_equal(result_temp, expected_temp):    # tolerance for function slowness
    difference = expected_temp - result_temp
    return 0 <= difference <= 2


class TestTeahouse(unittest.TestCase):

    def setUp(self):
        self.teacup = CoffeeMachine.brew('tea')
        self.coffee_cup = CoffeeMachine.brew('coffee')
        self.bread_cup = CoffeeMachine.brew('bread')

    def test_has_attribute(self):
        self.assertTrue(hasattr(self.teacup, 'full'))
        self.assertFalse(hasattr(self.teacup, 'fill'))

    def test_wrong_drink(self):
        self.assertIsInstance(self.bread_cup, str)
        self.assertFalse(hasattr(self.bread_cup, 'price'))

    def test_right_drink(self):
        self.assertIsInstance(self.teacup, object)
        self.assertTrue(hasattr(self.teacup, 'price'))
        self.assertEqual(self.teacup.price, 1.80)
        
    def test_drink(self):
        self.teacup.drink(0.2)
        self.assertRaises(Exception, self.teacup.drink, -2)
        self.assertEqual(0.8, self.teacup.full)

    def test_heat_up(self):
        self.assertRaises(Exception, self.teacup.heat_up, -20)
        self.assertTrue(self.teacup.temp < 80)
        temp = self.teacup.temp
        self.teacup.heat_up(10)
        self.assertTrue(almost_equal(self.teacup.temp, temp + 10))    

    def test_has_timestamp(self):
        self.assertTrue(hasattr(self.teacup, '_timestamp'))

    def test_cool_down(self):    
        self.assertEqual(self.coffee_cup._init_temp, 80)
        self.coffee_cup._timestamp -= 1200    # set back start time by 20 minutes
        start_temp = self.coffee_cup.temp
        self.coffee_cup.heat_up(20)
        self.assertTrue(almost_equal(self.coffee_cup.temp, start_temp + 20))

    def tearDown(self):
        pass

if __name__ == '__main__':
    unittest.main()


mymodule.py
import time
import math


class Cup(object):
    def __init__(self, drink, price):
        self.content = drink
        self.price = price
        self._init_temp = 80
        self.full = 1.0
        self._timestamp = time.time()    # initiate time for cool-down

    def heat_up(self, degrees):
        if degrees < 0:
            raise Exception('heat_up accepts only positive numbers')
        new_temp = self.temp + degrees
        if new_temp <= 100:
            self._init_temp = new_temp
        else:
            self._init_temp = 100
        self._timestamp = time.time()    # refresh time for cool-down

    def drink(self, sip):
        if sip < 0:
            raise Exception('drink only accepts positive numbers')
        self.full -= sip
        if self.full <= 0:
            self.full = 0

    @property
    def temp(self):
        exponent = time.time() - self._timestamp
        return int(self._init_temp * math.exp(-0.0002 * exponent))
        

class CoffeeMachine(object):

    @staticmethod
    def brew(drink):
        price_list = {
            'tea': 1.80,
            'coffee': 2.10,
            'latte': 2.60,
            'chocolate': 2.10,
            'chai': 2.40
        }
        if drink in price_list:
            return Cup(drink, price_list[drink])
        else:
            return 'Drink not available.'

teahouse/teahouse.py
# -*- coding: utf-8 -*-
import time
import math


class Cup(object):
    def __init__(self, drink, price):
        self.content = drink
        self.price = price
        self._init_temp = 80
        self.full = 1.0
        self._timestamp = time.time()    # initiate time for cool-down

    def heat_up(self, degrees):
        if degrees < 0:
            raise Exception('heat_up accepts only positive numbers')
        new_temp = self.temp + degrees
        if new_temp <= 100:
            self._init_temp = new_temp
        else:
            self._init_temp = 100
        self._timestamp = time.time()    # refresh time for cool-down

    def drink(self, sip):
        if sip < 0:
            raise Exception('drink only accepts positive numbers')
        self.full -= sip
        if self.full <= 0:
            self.full = 0

    @property
    def temp(self):
        exponent = time.time() - self._timestamp
        return int(self._init_temp * math.exp(-0.0002 * exponent))
        

class CoffeeMachine(object):

    @staticmethod
    def brew(drink):
        price_list = {
            'tea': 1.80,
            'coffee': 2.10,
            'latte': 2.60,
            'chocolate': 2.10,
            'chai': 2.40
        }
        if drink in price_list:
            return Cup(drink, price_list[drink])
        else:
            return 'Drink not available.'